src/codegen.cpp

```diff
@@ -1250,6 +1250,30 @@ static const auto sync_gc_total_bytes_func = new JuliaFunction<>{
             {getInt64Ty(C)}, false); },
     nullptr,
 };
+static const auto jl_allocgenericmemory = new JuliaFunction<TypeFnContextAndSizeT>{
+    XSTR(jl_alloc_genericmemory),
+    [](LLVMContext &C, Type *T_Size) {
+        auto T_prjlvalue = JuliaType::get_prjlvalue_ty(C);
+        return FunctionType::get(T_prjlvalue, // new Memory
+                                {T_prjlvalue, // type
+                                T_Size        // nelements
+                                }, false); },
+        [](LLVMContext &C) {
+            AttrBuilder FnAttrs(C);
+            AttrBuilder RetAttrs(C);
+#if JL_LLVM_VERSION >= 160000
+            FnAttrs.addMemoryAttr(MemoryEffects::inaccessibleMemOnly(ModRefInfo::ModRef) | argMemOnly(MemoryEffects::ModRefInfo::Ref));
+#endif
+            FnAttrs.addAttribute(Attribute::WillReturn);
+            RetAttrs.addAlignmentAttr(Align(16));
+            RetAttrs.addAttribute(Attribute::NonNull);
+            RetAttrs.addAttribute(Attribute::NoAlias);
```

Comment:
Technically it does alias if `nelements == 0`. Is this really necessary? (we generally annotate the loads from it with tbaa_const, so that llvm won't need to care)

Reply:
I guess the noalias doesn't add much here.

Reply:
It would be more profitable if we could annotate noalias on a new array on the data pointer itself, though I'm not sure how best to do data.

Reply:
I think `load` has a slightly more general form of that, which is a noalias set derived from that load. Though we would probably need some custom pass to figure out which loads could be annotated, to make sure all loads from the same pointer get the same noalias scope